### ISSUE TITLE
[🐛] Messaging - Require Cycle

### ISSUE THREAD

**Issue by madmeatballs** [invertase/react-native-firebase#6897]
## Issue

Describe your issue here

Getting require cycle warning on @react-native-firebase/messaging

```
 WARN  Require cycle: node_modules/@react-native-firebase/messaging/lib/index.js -> node_modules/@react-native-firebase/messaging/modular/index.js -> node_modules/@react-native-firebase/messaging/lib/index.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
```

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
{
  "name": "AppName",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint .",
    "postinstall": "patch-package",
    "android:clean": "cd android && ./gradlew clean && cd ..",
    "android:apk": "cd android && ./gradlew assembleRelease && cd ..",
    "android:bundle": "cd android && ./gradlew bundleRelease && cd ..",
    "ios:install": "cd ios && pod install && cd .."
  },
  "dependencies": {
    "@apollo/client": "^3.6.9",
    "@eva-design/eva": "^2.1.1",
    "@gorhom/bottom-sheet": "^4",
    "@notifee/react-native": "^7.3.0",
    "@openspacelabs/react-native-zoomable-view": "^2.1.1",
    "@react-native-clipboard/clipboard": "^1.11.1",
    "@react-native-community/datetimepicker": "^6.2.0",
    "@react-native-community/netinfo": "^9.3.7",
    "@react-native-firebase/analytics": "^17.0.0",
    "@react-native-firebase/app": "^17.0.0",
    "@react-native-firebase/crashlytics": "^17.0.0",
    "@react-native-firebase/messaging": "^17.0.0",
    "@react-native-masked-view/masked-view": "^0.2.6",
    "@react-native-picker/picker": "^2.4.1",
    "@react-navigation/bottom-tabs": "^6.3.1",
    "@react-navigation/devtools": "^6.0.11",
    "@react-navigation/native": "^6.0.10",
    "@react-navigation/native-stack": "^6.6.2",
    "@shopify/flash-list": "^1.4.0",
    "@tanstack/react-query": "^4.16.1",
    "@ui-kitten/components": "^5.1.2",
    "apollo3-cache-persist": "^0.14.1",
    "axios": "^0.27.2",
    "dayjs": "^1.11.5",
    "graphql": "^16.5.0",
    "libphonenumber-js": "^1.10.18",
    "lodash": "^4.17.21",
    "lottie-ios": "3.4.0",
    "lottie-react-native": "^5.1.4",
    "native-base": "3.4.22",
    "payment": "^2.4.6",
    "react": "17.0.2",
    "react-hook-form": "^7.36.0",
    "react-native": "0.68.6",
    "react-native-collapsible": "^1.6.0",
    "react-native-config": "1.5.0",
    "react-native-confirmation-code-field": "^7.3.0",
    "react-native-device-info": "^10.3.0",
    "react-native-fast-image": "^8.6.1",
    "react-native-flash-message": "^0.2.1",
    "react-native-gesture-handler": "^2.4.2",
    "react-native-image-gallery": "^2.1.5",
    "react-native-inappbrowser-reborn": "^3.7.0",
    "react-native-linear-gradient": "^2.6.2",
    "react-native-lottie-splash-screen": "^1.0.1",
    "react-native-mmkv": "2.5.1",
    "react-native-mmkv-flipper-plugin": "^1.0.0",
    "react-native-modalize": "^2.0.13",
    "react-native-onboarding-swiper": "^1.2.0",
    "react-native-phone-number-input": "^2.1.0",
    "react-native-reanimated": "2.6.0",
    "react-native-reanimated-carousel": "^2.5.0",
    "react-native-render-html": "^6.3.4",
    "react-native-responsive-dimensions": "^3.1.1",
    "react-native-safe-area-context": "^4.2.5",
    "react-native-scalable-image": "^1.1.0",
    "react-native-screens": "3.14.1",
    "react-native-share": "^8.0.0",
    "react-native-super-grid": "^4.4.0",
    "react-native-svg": "13.7.0",
    "react-native-swipe-list-view": "^3.2.9",
    "react-native-ui-lib": "^6.17.1",
    "react-native-vector-icons": "^9.1.0",
    "react-phone-number-input": "^3.2.3",
    "traverse": "^0.6.6",
    "zustand": "4.1.2"
  },
  "devDependencies": {
    "@babel/core": "^7.12.9",
    "@babel/runtime": "^7.12.5",
    "@react-native-community/eslint-config": "^2.0.0",
    "babel-jest": "^26.6.3",
    "babel-plugin-module-resolver": "^4.1.0",
    "deprecated-react-native-prop-types": "^2.3.0",
    "eslint": "^7.32.0",
    "jest": "^26.6.3",
    "metro-react-native-babel-preset": "^0.67.0",
    "patch-package": "^6.5.0",
    "react-native-flipper": "0.162.0",
    "react-native-flipper-zustand": "^2.0.1",
    "react-native-svg-transformer": "^1.0.0",
    "react-test-renderer": "17.0.2"
  },
  "jest": {
    "preset": "react-native"
  }
}

```

#### `firebase.json` for react-native-firebase v6:

```json
{
    "react-native": {
      "crashlytics_debug_enabled": true,
      "crashlytics_disable_auto_disabler": true,
      "crashlytics_auto_collection_enabled": true,
      "crashlytics_is_error_generation_on_js_crash_enabled": true,
      "crashlytics_javascript_exception_handler_chaining_enabled": true
    }
}
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

platform :ios, '11.0'
install! 'cocoapods', :deterministic_uuids => false

target 'AppName' do
  config = use_native_modules!

  # Flags change depending on the env values.
  flags = get_default_flags()

  pod 'Firebase', :modular_headers => true
  pod 'FirebaseCore', :modular_headers => true
  pod 'GoogleUtilities', :modular_headers => true
  $RNFirebaseAsStaticFramework = true

  use_react_native!(
    :path => config[:reactNativePath],
    # to enable hermes on iOS, change `false` to `true` and then install pods
    # :hermes_enabled => flags[:hermes_enabled],
    :hermes_enabled => true,
    :fabric_enabled => flags[:fabric_enabled],
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )





  target 'AppNameTests' do
    inherit! :complete
    # Pods for testing
  end

  # Enables Flipper.
  #
  # Note that if you have use_frameworks! enabled, Flipper will not work and
  # you should disable the next line.
  use_flipper!()

  post_install do |installer|
    react_native_post_install(installer)
    __apply_Xcode_12_5_M1_post_install_workaround(installer)
    # Add these lines for Xcode 14 builds
    installer.generated_projects.each do |project|
      project.targets.each do |target|
          target.build_configurations.each do |config|
              config.build_settings["DEVELOPMENT_TEAM"] = "UF2FQ56Q4Q"
          end
      end
    end
    # End of added lines
  end
end

```

#### `AppDelegate.m`:

```objc
#import "AppDelegate.h"

#import <React/RCTLinkingManager.h>
#import <Firebase.h>
#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>
#import "RNSplashScreen.h" // here

#import "AppName-Swift.h" // here, change project name to yours

#import <React/RCTAppSetupUtils.h>

#if RCT_NEW_ARCH_ENABLED
#import <React/CoreModulesPlugins.h>
#import <React/RCTCxxBridgeDelegate.h>
#import <React/RCTFabricSurfaceHostingProxyRootView.h>
#import <React/RCTSurfacePresenter.h>
#import <React/RCTSurfacePresenterBridgeAdapter.h>
#import <ReactCommon/RCTTurboModuleManager.h>

#import <react/config/ReactNativeConfig.h>

@interface AppDelegate () <RCTCxxBridgeDelegate, RCTTurboModuleManagerDelegate> {
  RCTTurboModuleManager *_turboModuleManager;
  RCTSurfacePresenterBridgeAdapter *_bridgeAdapter;
  std::shared_ptr<const facebook::react::ReactNativeConfig> _reactNativeConfig;
  facebook::react::ContextContainer::Shared _contextContainer;
}
@end
#endif

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  RCTAppSetupPrepareApp(application);

//  rnfirebase
  [FIRApp configure];
  
  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];

#if RCT_NEW_ARCH_ENABLED
  _contextContainer = std::make_shared<facebook::react::ContextContainer const>();
  _reactNativeConfig = std::make_shared<facebook::react::EmptyReactNativeConfig const>();
  _contextContainer->insert("ReactNativeConfig", _reactNativeConfig);
  _bridgeAdapter = [[RCTSurfacePresenterBridgeAdapter alloc] initWithBridge:bridge contextContainer:_contextContainer];
  bridge.surfacePresenter = _bridgeAdapter.surfacePresenter;
#endif

  UIView *rootView = RCTAppSetupDefaultRootView(bridge, @"AppName", nil);

  if (@available(iOS 13.0, *)) {
    rootView.backgroundColor = [UIColor systemBackgroundColor];
  } else {
    rootView.backgroundColor = [UIColor whiteColor];
  }

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];

  //set rootView background color
  rootView.backgroundColor = [UIColor colorWithRed: 1.00 green: 0.94 blue: 0.94 alpha: 1.00]; // change backgroundColor

  Dynamic *t = [Dynamic new];
  UIView *animationUIView = (UIView *)[t createAnimationViewWithRootView:rootView lottieName:@"logo_animated"]; // change lottieName to your lottie files name

  //instead of rootView, create a full screen UIView called animationContainerView with the same background color as the rootView
  UIView *animationContainerView = [[UIView alloc] initWithFrame:rootView.frame];
  animationContainerView.backgroundColor = [UIColor colorWithRed: 1.00 green: 0.94 blue: 0.94 alpha: 1.00]; // change backgroundColor

  //add animationUIView to animationContainerView
  [animationContainerView addSubview:animationUIView];

  //register animationContainerView to RNSplashScreen
  [RNSplashScreen showLottieSplash:animationContainerView inRootView:rootView];

  // casting UIView type to AnimationView type
  AnimationView *animationView = (AnimationView *) animationUIView;

  // play
  [t playWithAnimationView:animationView];

  // If you want the animation layout to be forced to remove when hide is called, use this code
  [RNSplashScreen setAnimationFinished:true];
  
  return YES;
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index"];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

#if RCT_NEW_ARCH_ENABLED

#pragma mark - RCTCxxBridgeDelegate

- (std::unique_ptr<facebook::react::JSExecutorFactory>)jsExecutorFactoryForBridge:(RCTBridge *)bridge
{
  _turboModuleManager = [[RCTTurboModuleManager alloc] initWithBridge:bridge
                                                             delegate:self
                                                            jsInvoker:bridge.jsCallInvoker];
  return RCTAppSetupDefaultJsExecutorFactory(bridge, _turboModuleManager);
}

#pragma mark RCTTurboModuleManagerDelegate

- (Class)getModuleClassFromName:(const char *)name
{
  return RCTCoreModulesClassProvider(name);
}

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:(const std::string &)name
                                                      jsInvoker:(std::shared_ptr<facebook::react::CallInvoker>)jsInvoker
{
  return nullptr;
}

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:(const std::string &)name
                                                     initParams:
                                                         (const facebook::react::ObjCTurboModule::InitParams &)params
{
  return nullptr;
}

- (id<RCTTurboModule>)getModuleInstanceFromClass:(Class)moduleClass
{
  return RCTAppSetupDefaultModuleFromClass(moduleClass);
}

#endif

- (BOOL)application:(UIApplication *)application
   openURL:(NSURL *)url
   options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
{
  return [RCTLinkingManager application:application openURL:url options:options];
}

@end

```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
System:
    OS: macOS 13.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 91.50 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 19.5.0 - /opt/homebrew/bin/node
    Yarn: 1.22.19 - ~/node_modules/.bin/yarn
    npm: 9.3.1 - /opt/homebrew/bin/npm
    Watchman: Not Found
  Managers:
    CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.2, iOS 16.2, macOS 13.1, tvOS 16.1, watchOS 9.1
    Android SDK:
      API Levels: 30, 31, 33
      Build Tools: 30.0.2, 30.0.3, 31.0.0, 33.0.0, 33.0.1
      System Images: android-21 | Google APIs ARM 64 v8a, android-23 | Google APIs ARM 64 v8a, android-28 | Google ARM64-V8a Play ARM 64 v8a, android-29 | Google APIs ARM 64 v8a, android-29 | Google Play ARM 64 v8a, android-30 | Google Play ARM 64 v8a, android-31 | Google APIs ARM 64 v8a, android-31 | Google Play ARM 64 v8a, android-33 | Google APIs ARM 64 v8a
      Android NDK: Not Found
  IDEs:
    Android Studio: 2022.1 AI-221.6008.13.2211.9477386
    Xcode: 14.2/14C18 - /usr/bin/xcodebuild
  Languages:
    Java: 18.0.1 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2 
    react-native: 0.68.6 => 0.68.6 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [x] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `17.0.0`
- **`Firebase` module(s) you're using that has the issue:**
  - `Messaging`
- **Are you using `TypeScript`?**
  - `N` & `N/A`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by jonathanm-tkf** [comment]
Same here

**Comment by mikehardy** [comment]
No argument - this cycle does exist, yet it is a warning without functional consequence, so it has not gotten priority in the past and may not in the future unless a motivated person posts a PR. A PR would be welcome though!

**Comment by tomwanzek** [comment]
Same require cycle issue also exists for Firebase Analytics

```
Require cycle: node_modules/@react-native-firebase/analytics/lib/index.js -> node_modules/@react-native-firebase/analytics/modular/index.js -> node_modules/@react-native-firebase/analytics/lib/index.js
```

So seems it's a general code organization pattern which would need changing across the Firebase modules. I.e. the cyclical import path a -> b -> a needs to be broken by moving shared code objects into a file c s.t. c -> a and c -> b. As usual in these kind of scenarios.

Not sure, if I get to looking at the details to create a PR in the short run. It seems, the past few weeks have just been driven more by issues creeping in from 3rd party libs than doing planned feature work on our project 😅

**Comment by mikehardy** [comment]
> Not sure, if I get to looking at the details to create a PR in the short run. It seems, the past few weeks have just been driven more by issues creeping in from 3rd party libs than doing planned feature work on our project 

:laughing: that's how I ended up becoming a maintainer here

**Comment by tomwanzek** [comment]
@mikehardy I hear your nudge-nudge-wink-wink 😄 . And I really appreciate all the hard work you are putting in 🥇

**Comment by tomwanzek** [comment]
@mikehardy I opened a draft PR https://github.com/invertase/react-native-firebase/pull/6940 with a possible fix for the cyclical import dependency.

My limitation right now is two-fold:
1) I'm not familiar with the modularization approach/next steps RN Firebase is undergoing right now to know if there is any architectural reason, why the creation of the Firebase root instance constant cannot be moved into the `/modular/index.js` file. I.e. ensuring that all imports are always directed from `/modular` to `/lib`. Or if an alternative approach would be needed were the root instance creation and export should be done in a separate file, possibly co-located with the `lib/index.js` file and both `/modular/index.js` and `/lib/index.js` import from that file to break the cycle.
2)  In both the Analytics and Messaging current implementation, there are comments left in the code. I'm not cure if that means there was still work to be done which affects the proposed fix, or if those are obsolete legacy comments which can be removed. See:
   * https://github.com/invertase/react-native-firebase/blob/main/packages/analytics/lib/index.js#L785-L788
   * https://github.com/invertase/react-native-firebase/blob/main/packages/messaging/lib/index.js#L541-L544

Not directly related, I noticed that the `Podfile.lock` versions of the Firebase related pods are not current. I.e. running `yarn tests:ios:pod:install` will create a diff between versions `17.1.0` to `17.3.1`. Is there any reason, why that diff was not previously committed? I discarded it in my draft PR for the time being 🤔 

Thanks in advance for your thoughts on it 🙇 

I'll cross-reference this comment on the PR for consideration.

**Comment by tomwanzek** [comment]
PS: If the suggested fix for Analytics & Messaging is indeed the most appropriate one, than I suppose the same pattern should be applied in the modularization of any of the other RN Firebase packages.

**Comment by tomwanzek** [comment]
Any update on reviewing the PR #6940 I opened?

**Comment by dmgcoding** [comment]
will downgrading firebase messaging solve this???

**Comment by mikehardy** [comment]
@dmgcoding "solve" in the sense of resolving a known issue? As far as I understand, it is a *warning* and I'm unaware of anyone demonstrating it causes any actual issues that would be resolved. Breaking the require cycle would eliminate warnings, and the *possibility* of issues, but - no one has demonstrated it is actually causing a problem.

Downgrading means specifically you will be using older software where you would *definitely* then be subject to actual issues that have resolutions in current software.

So rationally, I can't recommend downgrading software in order to quiet a warning

**Comment by tomwanzek** [comment]
Any update on next steps for PR #6940? Just pinging here, so the issue does not get stale-marked for no activity 😄

**Comment by tomwanzek** [comment]
Please remove the stale label from the issue, there is an open PR with some open questions I was hoping to get guidance on: https://github.com/invertase/react-native-firebase/pull/6940#issuecomment-1554634742 cc @mikehardy @Lyokone @russellwheatley

**Comment by tomwanzek** [comment]
Closed my PR for this issue for now, as unfortunately I could not get any response to my follow-up questions and both PR and this issue keep getting stale marked as a result. Feel free to revisit the suggestions contained as a fix in PR #6940 , would have been nice to get this fixed 😢 across the board as outlined in my questions.
